### PR TITLE
fix a deadlock due to getting segmentlock before snapshot lock when replacing segment

### PR DESCRIPTION
**Pull request by klsince**
This was to refine https://github.com/apache/pinot/pull/12105

We can't just reuse the SegmentLocks in RealtimeSegmentManager.handleUpsert() to synchronize the segment replacement logic. Because many methods of partitionUpsertMetadataManager takes this lock internally, but after taking snapshot RW lock. If we take segmentLock here (before taking snapshot RW lock), we can get into deadlock with the other threads calling other methods of partitionUpsertMetadataManager, like removeSegment.

So this PR added a private UpsertSegmentLocks just to sync segment replacement inside handleUpsert() method.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java [hunk 1]
```diff
@@ -562,6 +570,25 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
     }
   }
 
+  // Same as the SegmentLocks, but just for handleUpsert method to synchronize threads doing segment replacement.
+  private static class UpsertSegmentLocks {
```
Comment (walterddr) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: We can make the class abstract and create 2 static instance one for generic and one for upsert. (This way we can extend more later)

**Review by walterddr — COMMENTED** [review]
Lgtm mostly. Just one comment on the lock impl. Ptal

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java (hunk 1 above)
Comment (klsince) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: But all members in this lock util class are static. So I assume no need to define an abstract base class in this case.

Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: i was thinking: 
```
public class SegmentLocks {
  private static final SegmentLocks GENERIC_LOCK = new SegmentLocks();
  private static final SegmentLocks UPSERT_LOCK = new SegmentLocks();

  private final Lock[] _locks = new Lock[NUM_LOCKS];

  private SegmentLocks() {
    for (int i = 0; i < NUM_LOCKS; i++) {
      _locks[i] = new ReentrantLock();
    }
  }

  private Lock getLock(String tableNameWithType, String segmentName) {
    return _locks[Math.abs((31 * tableNameWithType.hashCode() + segmentName.hashCode()) % NUM_LOCKS)];
  }

  public static Lock getSegmentLock(String tableNameWithType, String segmentName) {
    return GENERIC_LOCK.getLock(tableNameWithType, segmentName);
  }

  public static Lock getUpsertLock(String tableNameWithType, String segmentName) {
    return UPSERT_LOCK.getLock(tableNameWithType, segmentName);
  }
```

**Review by walterddr — APPROVED** [review]
approved

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java (hunk 1 above)
Comment (klsince) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: Not bad

Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: this way we can extend more if we have to (just add one more static member). sorry for the "abstract" comment, i realized that it was a static class afterwards

Comment (klsince) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:574]: Your change is neat. But on a second thought, I kinda prefer to keep this upsert lock util class private to RealtimeDataManager for now, until we find other classes that also need to sync some upsert logic

**Review by walterddr — APPROVED** [review]
🚀

**Comment by Jackie-Jiang** [comment]
We might need to revisit this fix. Using a separate lock cannot protect the replaced segment being removed at the same time

**Pull request merged**